### PR TITLE
GLTFExporter image caching.

### LINKED ISSUE #14063: GLTFExporter caches textures not images
The current implementation of GLTFExporter is caching textures as it processes the scene but other parts of three js (e.g. TextureLoader and ImageLoader) cache the image.

This creates problems when, for instance, you load various model files (via FBXLoader etc) that have references to the same image files and you try to export the resulting scene via GLTFExporter with embedImages = true and binary = true. Essentially, it packs duplicate image data into the glb, which bloats it considerably.

The problem seems to centre around the fact that TextureLoader creates a new texture (with a potentially cached Image) every time you call load() but GLTFExporter is basing it's image caching on the texture when you try to process the scene. You end up with a list of textures all referring to the same image but GLTFExporter will go through each texture and write out the corresponding cached image.

NOTE: I could solve this by hacking in something to patch/reformat all the textures in my scene to work with GLTFExporter but I think handling this in the exporter would be better and more user-friendly.

##### Three.js version

- [ ] Dev
- [x ] r92
- [ ] ...

##### Browser

- [x] All of them
- [ ] Chrome
- [ ] Firefox
- [ ] Internet Explorer

##### OS

- [x] All of them
- [ ] Windows
- [ ] macOS
- [ ] Linux
- [ ] Android
- [ ] iOS

### PR DESCRIPTION
**Pull request by kickblade**
@takahirox @donmccurdy 

I've added a map for images that prevents multiple writes of the same data. I followed the same pattern as the other cache mechanisms in there. It seems to actually be a very simple change to get it working, let me know if I've missed something though.

This reduces output glb size considerably for me when the scene contains shared images and embedImages is enabled for the export. My test scene went down from ~10Mb to ~1.2Mb. I've also checked the cache works when there is a mix of shared and unique textures to make sure it handles those cases correctly as well.

closes #14063

### CODE REVIEW

File: examples/js/exporters/GLTFExporter.js
Comment (takahirox) [inline examples/js/exporters/GLTFExporter.js]: You need an empty line after `return index;` for the code style.

**Comment by takahirox** [comment]
Thanks.

I think we should remove `Texture` specific code from `processImage()`.

https://github.com/kickblade/three.js/blob/5e50c2138dc6755bc5df4aa7b1c9bb63bbdcf0af/examples/js/exporters/GLTFExporter.js#L640-L645

If two textures share one image and one texture's `.flipY` is `true` while another one is `false`, 

https://github.com/kickblade/three.js/blob/5e50c2138dc6755bc5df4aa7b1c9bb63bbdcf0af/examples/js/exporters/GLTFExporter.js#L619

similarly, if one texutre's `.format` is `RGBAFormat` while another one is not, texture will be broken.

**Comment by donmccurdy** [comment]
> If two textures share one image and one texture's .flipY is true while another one is false...

> ...if one texutre's .format is RGBAFormat while another one is not, one of the textures will be broken.

Hm, yeah both of these cases complicate things. 😕 Maybe the cache should store a list of entries, like this?

```js
if ( cachedData.images.has( map.image ) ) {
  
  var imageCacheItems = cachedData.images.get( map.image );

  for ( var i = 0; i < imageCacheItems.length; ++ i ) {

    var imageCacheItem = imageCacheItems[ i ];
    if ( imageCacheItem.flipY !== map.flipY ) continue;
    if ( imageCacheItem.format !== map.format ) continue;

    return imageCacheItem.image;

  }

  var image = cachedImages.find( function ( image ) {

  } );

}
```

Where `imageCacheItem` retains data like `{format: ..., flipY: ...; image: ...}`.

**Comment by donmccurdy** [comment]
There is no `flipY` in glTF (maybe it was in a previous version?) but the `KHR_texture_transforms` extension will add offset/repeat/rotation. I think duplicating the image if it's used with flipY both true and false is fine, that should be rare. When texture transforms are supported (eventually) we should try to avoid duplicating the image though.

**Comment by takahirox** [comment]
Yeah, I think duplication is ok so far, too.

> we should try to avoid duplicating the image though.

by simulating .flipY with offset + repeat of `KHR_texture_transforms` extension, correct?

**Comment by takahirox** [comment]
Hm, I've misread?

> When texture transforms are supported 

That meant, "When `flipY` flag is supported in glTF texture spec"?

**Comment by donmccurdy** [comment]
> That meant, "When flipY flag is supported in glTF texture spec"?

No, I don't think that `flipY` will be supported in the glTF spec. glTF _will_ eventually have texture UV transforms (offset/repeat/rotation), but they are not really the same thing as `texture.flipY`, so I don't think we should try to use one to imitate the other. So if one image is used with two different values of `flipY`, we should either duplicate the image or throw an error. It will be too complicated otherwise — what about when texture scale is animated, and then we have to rewrite the animation values to reflect `flipY`? I don't think that's worth it, too much work for a very unlikely case.

**Comment by takahirox** [comment]
I have the same opinion. I don't think it's a good idea to simulate `.flipY` with transforms.

What I wanted to first do is to understand what you mentioned. Probably this is just a English question. (Sorry for my bad English!) 

> When texture transforms are supported (eventually) we should try to avoid duplicating the image though.

I thought you mentioned that "When `KHR_texture_transforms` extension becomes available we Three.js glTF exporter can avoid exporting two images for flipY true and false by using the extension" but it seems I'm wrong?

**Comment by donmccurdy** [comment]
> I thought you mentioned that "When KHR_texture_transforms extension becomes available we Three.js glTF exporter can avoid exporting two images for flipY true and false by using the extension" but it seems I'm wrong?

Oops sorry, I wrote that poorly... `KHR_texture_transforms` should let us export `texture.offset` and `texture.repeat` but should not have any relationship to `texture.flipY`. We would keep duplicating the image if two `flipY` values are needed, even with `KHR_texture_transforms`.

**Comment by takahirox** [comment]
OK OK, I understand. As I said, I have the same opinion. Thanks for patience!

**Comment by kickblade** [comment]
These are all great points guys. I'll update the empty line, easy. Will try a few ideas out and get back to you soon, bit snowed at the moment. Current thoughts though:

- processImage(map) should be processImage(image, mimeType, flipY)?
Allowing access to map/texture-centric properties after the cache checks will be dangerous now. I think it makes sense to change this so we restrict it. It also makes the function a bit more predictable too (processImage processes images instead of textures). 
- Maybe there's a way to generate a map key from these params, so we can have a more direct lookup (as an alternative to pulling an array of different configs).

I'll have a muck around anyway and get back to you guys when I have something useful :-)

**Comment by donmccurdy** [comment]
I couldn't think of a way to obtain a key from `texture.image` without modifying the element itself (`image.src` is not necessarily safe; it could be a `<canvas/>`). One alternative to the array of configs would be a separate map of IDs I suppose:

```
imageElementKeys = Map()
if ( !imageElementKeys.has( image ) ) {
  imageElementKeys.set( image, s.size );
}
var imageKey = imageElementKeys.get( image ) + ':' + texture.flipY + ':' + texture.format;
if ( ! imageCache.has( imageKey ) {
  // ...
}
```

I'm fine with changing the signature of `processImage()`.

**Comment by kickblade** [comment]
Hi @donmccurdy, just to let you know I've got back on to trying a few things out for this. I think both of your ideas are good, just wouldn't mind trying some ideas out before we plump for one. Maybe give me until the end of the weekend and then you can make a call on the one to put in?

No wonder this feature was marked as a TODO :-)

**Comment by kickblade** [comment]
@donmccurdy @takahirox 

I've had a good read around and since Map() only supports reference-types for keys (I'm a C++ dev btw so this is all quite new to me) there doesn't appear to be a way to generate a key without adding a lot of complexity or using an external library (such as blockhash-js, crypto-js etc). I did come across a couple of forums discussing ES7 value-type implementation for Map() mentioned here: https://esdiscuss.org/topic/value-types-as-map-keys-in-es7 but I've searched all over the place and can't find very much else on it.

So I think for this upcoming milestone, it comes down to a choice between one of your two Map() ideas or an alternative Array implementation:

cachedData.images = []
...
function processImage( image, mimeType, flipY ) {
	var result = cachedData.images.find( function( entry ) {
		return 	entry.img === image &&
				entry.mimeType === mimeType &&
				entry.flipY === flipY;
	});

	if( result != undefined ) {
		return result.index;
	}
...
	var entry = {
		img: image,
		mimeType: mimeType,
		flipY: flipY,
		index: outputJSON.images.length - 1
	}
	cachedData.images.push( entry );
	return entry.index;
}

O(n) worst case for the array versus ~O(log n) worst case for the maps. The advantage of the array is that it's simple (just the one datastructure to navigate). Sorry it's taken me so long, happy to go with whatever you guys decide anyway :-)

**Comment by takahirox** [comment]
Thanks for sharing the idea.

Personally I prefer scalable algorithm. How about `Map` + `Object`?

    if ( ! cachedData.images.has( map.image ) ) cachedData.images.set( map.image, {} );

    var cachedImages = cachedData.images.get( map.image );

    var key = map.format;
    if ( map.flipY ) key += ':flip';

    if ( cachedImages[ key ] !== undefined ) return cachedImages[ key ];

    ....

    outputJSON.images.push( gltfImage );

    var index = outputJSON.images.length - 1;
    cachedImages[ key ] = index;

    return index;

**Comment by kickblade** [comment]
@takahirox yeh I think that's the neatest solution, nice! I've updated the pull request so you can hopefully the see the final version. I made one minor change to the way the key was constructed to get rid of the 'if' statement. Personally think it's clearer to construct that key on one line with a standard formation but let me know if you disagree and I'll update it no worries at all :-) I've tested this with my particular use-case and it works gloriously! Let me know if there's anything else in there you'd rather change too, it's all good :-)

**Comment by takahirox** [comment]
Thanks for the update!

File: examples/js/exporters/GLTFExporter.js
Comment (takahirox) [inline examples/js/exporters/GLTFExporter.js]: I prefer moving this line into `processImage()`.

**Comment by kickblade** [comment]
Sweet, I'll do this over the weekend. Makes sense.

**Comment by kickblade** [comment]
Just pushed an update to the branch with that change @takahirox. Also added "flipY/" explicitly in the key so now it reads "image/png:flipY/true" or "image/jpg:flipY/false" etc. @donmccurdy can you confirm this commit addresses the change you were referring to, wasn't sure if it was the one takahirox requested or the one I made to the key :-)

Thank you very much for all your help and input guys! This is my first contribution to a public github project, it's been a pleasure. As always, let me know if there's anything you need me to do before merging it in.

**Review by donmccurdy — APPROVED** [review]

**Comment by donmccurdy** [comment]
> @donmccurdy can you confirm this commit addresses the change you were referring to...

Yep, looks good. Thank you!

File: examples/js/exporters/GLTFExporter.js
Comment (takahirox) [inline examples/js/exporters/GLTFExporter.js]: nit: `flipY`. You've dropped `Y`.

**Comment by takahirox** [comment]
Nice work! I request one minor comment correction but others look fine.

**Comment by kickblade** [comment]
Done :-) do you need me to squash any of these commits or will you just merge the whole branch history over?

**Review by takahirox — APPROVED** [review]

**Comment by takahirox** [comment]
It's up to @mrdoob but IMO I don't think you need to squash

**Comment by donmccurdy** [comment]
@mrdoob this is ready to merge. 👍

**Comment by mrdoob** [comment]
Thanks!

**Pull request merged**
